scipy/stats/_morestats.py

```diff
@@ -969,13 +975,9 @@ def boxcox_llf(lmb, data):
         # lead to loss of precision.
         # Division by lmb can be factored out to enhance numerical stability.
         logx = lmb * logdata
-        # convert to `np` for `special.logsumexp`
-        logx = np.asarray(logx)
-        logvar = _log_var(logx) - 2 * np.log(abs(lmb))
-        logvar = xp.asarray(logvar)
+        logvar = _log_var(logx, xp) - 2 * math.log(abs(lmb))
 
-    res = (lmb - 1) * xp.sum(logdata, axis=0) - N/2 * logvar
-    res = xp.astype(res, dt)
+    res = (lmb - 1) * xp.sum(logdata, axis=0, dtype=dt) - N/2 * logvar
```

Comment:
Only need `dtype=dt` until 2023.12 rules kick in.

Reply:
But of course, we still have pre-NEP 50 promotion rules to worry about.
```suggestion
    res = (lmb - 1) * xp.sum(logdata, axis=0) - N/2 * logvar
    res = xp.astype(res, dt)
```
That's probably why we had it this way, so I'll leave it.